packages/flutter/lib/src/foundation/collections.dart

```diff
@@ -45,3 +47,79 @@ bool listEquals<T>(List<T> a, List<T> b) {
   }
   return true;
 }
+
+/// A map of key-value pairs where each key is a [Type] and its value is of
+/// that type.
+class TypedDictionary {
+  /// Creates an empty, mutable, typed dictionary.
+  TypedDictionary() : _storage = HashMap<Type, Object>();
+
+  const TypedDictionary._empty() : _storage = const <Type, Object>{};
+
+  /// Creates a typed dictionary whose values cannot be changed.
+  ///
+  /// The dictionary is initialized to the values of the given dictionary. The
+  /// other dictionary is copied; modifying it will not modify this dictionary.
+  ///
+  /// If the dictionary is large, this call may be expensive (O(N) in the number
+  /// of entries), since the values are eagerly copied.
+  TypedDictionary.unmodifiable(TypedDictionary other) :
+        assert(other != null),
+        _storage = Map<Type, Object>.unmodifiable(other._storage);
+
+  /// An empty, immutable, [TypedDictionary].
+  ///
+  /// This object is equivalent to the object obtained from
+  /// `TypedDictionary.unmodifiable(TypedDictionary())`, but the value is a
+  /// compile-time constant.
+  static const TypedDictionary empty = TypedDictionary._empty();
+
+  final Map<Type, Object> _storage;
+
+  /// Stores data in the dictionary under the key `T`.
+  ///
+  /// Any previously-existing data associated with that type is removed.
+  ///
+  /// An exception will be thrown if this dictionary was created using
+  /// [TypedDictionary.unmodifiable].
+  void set<T>(T data) {
+    _storage[T] = data;
```

Comment:
For future archeologists: This was suggested to avoid the following surprise:

```dart
var x;
x = 2;
var args = TypedDictionary()..set(x); 
...
args.get<int>();
> null
```